cluster/swarm/cluster.go

```diff
@@ -66,35 +66,16 @@ func (c *Cluster) Handle(e *cluster.Event) error {
 
 // CreateContainer aka schedule a brand new container into the cluster.
 func (c *Cluster) CreateContainer(config *dockerclient.ContainerConfig, name string) (*cluster.Container, error) {
+	c.scheduler.Lock()
+	defer c.scheduler.Unlock()
 
-	c.RLock()
-	defer c.RUnlock()
-retry:
-	// FIXME: to prevent a race, we check again after the pull if the node can still handle
-	// the container. We should store the state in the store before pulling and use this to check
-	// all the other container create, but, as we don't have a proper store yet, this temporary solution
-	// was chosen.
 	n, err := c.scheduler.SelectNodeForContainer(c.listNodes(), config)
 	if err != nil {
 		return nil, err
 	}
 
 	if nn, ok := n.(*node); ok {
-		container, err := nn.create(config, name, false)
```

Comment:
@smothiki It is a temporary fix. This will ensure that we can do heavy plumbing on the storage part and take scheduling decisions directly out of the store. I proposed a queue in another issue but basically between the `SelectNodeForContainer` and `addToQueue` part, we'll need to add a lot of checks (for name uniqueness, etc.). As this part is going to be temporary anyway, better to hold-on to the safest solution to work on the storage part in the meantime.